Only load from `cgi` what is required for Ruby 3.5 compatibility

Earlopain opened the pull request:
### Motivation / Background

In Ruby 3.5 most of the `cgi` gem will be removed (bundled gem). Only the various escape/unescape methods will be retained by default, which is luckily the only ones that rails actually uses.

Practically, requiring `cgi` will still work, it just emits a warning (and other things like `CGI::Cookie` obviously won't be available).

### Detail

https://bugs.ruby-lang.org/issues/21258

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [ ] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

yahonda commented:
I'm wondering if replacing all of `require "cgi"` with `require "cgi/escape"` is fine.

https://github.com/rails/rails/blob/747f85f200e7bb2c1a31b4e26e5a5655e2dc0cdc/actionpack/test/journey/router/utils_test.rb#L21-L23

At least `CGI.unescapeURIComponent` needs `require "cgi/util"`, Just `require "cgi/escape"` is not enough.

- require "cgi/escape" raises the NameError

```
$ ruby -v
ruby 3.4.3 (2025-04-14 revision d0b7e5b6a0) +PRISM [x86_64-linux]
$ gem info cgi

*** LOCAL GEMS ***

cgi (0.4.2)
    Author: Yukihiro Matsumoto
    Homepage: https://github.com/ruby/cgi
    Licenses: Ruby, BSD-2-Clause
    Installed at: /home/yahonda/.rbenv/versions/3.4.3/lib/ruby/gems/3.4.0

    Support for the Common Gateway Interface protocol.
$ irb
irb(main):001> require "cgi/escape"
=> true
irb(main):002> CGI.unescapeURIComponent("a%2Fb%20c+d")
(irb):2:in 'CGI::Escape#unescapeURIComponent': uninitialized class variable @@accept_charset in #<Class:CGI> (NameError)
	from (irb):2:in '<main>'
	from <internal:kernel>:168:in 'Kernel#loop'
	from /home/yahonda/.rbenv/versions/3.4.3/lib/ruby/gems/3.4.0/gems/irb-1.15.2/exe/irb:9:in '<top (required)>'
	from /home/yahonda/.rbenv/versions/3.4.3/bin/irb:25:in 'Kernel#load'
	from /home/yahonda/.rbenv/versions/3.4.3/bin/irb:25:in '<main>'
irb(main):003>
```

- require "cgi/util" is needed here
```
$ irb
irb(main):001> require "cgi/util"
=> true
irb(main):002> CGI.unescapeURIComponent("a%2Fb%20c+d")
=> "a/b c+d"
irb(main):003>
```

Related to https://github.com/rails/rails/pull/54517

Earlopain commented:
Oh, great catch. I was only looking if the method was defined. I will fix this

Earlopain commented:
I have added `require "cgi/util" if RUBY_VERSION < "3.5"`. I would like to solve it in a different way, but the fact that the methods are present (in a broken way) don't leave me with many ideas on how to figure that out at runtime in a different way.

byroot commented:
So I added the `cgi` dependency as a quick way to fix the issue: https://github.com/rails/rails/pull/55043

But I agree that when possible your solution is better. Feel free to rebase and remove the newly declared dependencies where applicable. (as you mentioned CGI::Cookie will still require a dependency)

Earlopain commented:
Right, so rails doesn't actually use any of the "advanced" cgi stuff. I just mentioned `CGI::Cookie` as an example. There are many downstream users that similarly do `require "cgi"`. For now, the only one I found that actually makes trouble here is `rack` (for which I have a PR at https://github.com/rack/rack/pull/2327). From all the gems that rails has in its lockfile, that actually seems to be the only one that really errors (the others just warn).

So here I pretty much reverted your PR and just added it to the gemfile instead. It will take long to fix every other gem (either changing the requires or adding it as a dependency when necessary) but for rails that isn't necessary.

byroot commented:
🤦 I went with adding the dependency because a quick grep showed matches for `CGI::Cookie`, but turns out it was in an old markdown file...

byroot commented on Gemfile:
```diff
@@ -45,6 +45,10 @@ gem "uri", ">= 0.13.1", require: false
 
+# Workaround until most gems handle Ruby 3.5+ cgi changes.
+# https://bugs.ruby-lang.org/issues/21258
+gem "cgi"
```
My only worry is that. How do we keep track of gems that need to be updated and make sure we remove this once we're done?

Earlopain replied:
Unless I missed one, only rack is actually using something else than escape/unescape. After that is fixed/released, you can move it to `StrictWarnings`, although that would only really capture a single one since the warning is only emitted once.

<details><summary>For reference, here is the list of matches</summary>
<p>

38 results - 38 files

vendor/ruby/3.4.0/gems/amq-protocol-2.3.2/lib/amq/uri.rb:
  3: require "cgi"

vendor/ruby/3.4.0/gems/aws-sdk-core-3.215.1/lib/aws-sdk-core/util.rb:
  3: require 'cgi'

vendor/ruby/3.4.0/gems/aws-sdk-core-3.215.1/lib/aws-sdk-core/endpoints/matchers.rb:
  3: require 'cgi'

vendor/ruby/3.4.0/gems/aws-sdk-core-3.215.1/lib/aws-sdk-core/xml/error_handler.rb:
  3: require 'cgi'

vendor/ruby/3.4.0/gems/aws-sdk-core-3.215.1/lib/seahorse/util.rb:
  3: require 'cgi'

vendor/ruby/3.4.0/gems/aws-sdk-core-3.215.1/lib/seahorse/client/net_http/connection_pool.rb:
  3: require 'cgi'

vendor/ruby/3.4.0/gems/aws-sdk-s3-1.177.0/lib/aws-sdk-s3/legacy_signer.rb:
  6: require 'cgi'

vendor/ruby/3.4.0/gems/aws-sdk-s3-1.177.0/lib/aws-sdk-s3/object_multipart_copier.rb:
  4: require 'cgi'

vendor/ruby/3.4.0/gems/aws-sdk-s3-1.177.0/lib/aws-sdk-s3/plugins/url_encoded_keys.rb:
  4: require 'cgi'

vendor/ruby/3.4.0/gems/aws-sigv4-1.11.0/lib/aws-sigv4/signer.rb:
  8: require 'cgi'

vendor/ruby/3.4.0/gems/azure-storage-common-2.0.4/lib/azure/core/auth/shared_key.rb:
  15: require 'cgi'

vendor/ruby/3.4.0/gems/azure-storage-common-2.0.4/lib/azure/storage/common/core/auth/shared_key.rb:
  26: require "cgi"

vendor/ruby/3.4.0/gems/brakeman-7.0.0/bundle/ruby/3.1.0/gems/erubis-2.7.0/lib/erubis/preprocessing.rb:
  6: require 'cgi'

vendor/ruby/3.4.0/gems/brakeman-7.0.0/lib/brakeman/messages.rb:
  89:     require 'cgi'

vendor/ruby/3.4.0/gems/brakeman-7.0.0/lib/brakeman/report/report_html.rb:
  1: require 'cgi'

vendor/ruby/3.4.0/gems/bundler-audit-0.9.2/lib/bundler/audit/cli/formats/junit.rb:
  19: require 'cgi'

vendor/ruby/3.4.0/gems/capybara-3.40.0/lib/capybara/rack_test/driver.rb:
  7: require 'cgi'

vendor/ruby/3.4.0/gems/faraday-1.10.4/lib/faraday.rb:
  3: require 'cgi'

vendor/ruby/3.4.0/gems/google-cloud-storage-1.54.0/lib/google/cloud/storage/file/signer_v2.rb:
  18: require "cgi"

vendor/ruby/3.4.0/gems/google-cloud-storage-1.54.0/lib/google/cloud/storage/file/signer_v4.rb:
  17: require "cgi"

vendor/ruby/3.4.0/gems/httpclient-2.9.0/sample/oauth_buzz.rb:
  38: require 'cgi'

vendor/ruby/3.4.0/gems/httpclient-2.9.0/test/helper.rb:
  9: require 'cgi'

vendor/ruby/3.4.0/gems/i18n-1.14.6/lib/i18n/exceptions.rb:
  3: require 'cgi'

vendor/ruby/3.4.0/gems/loofah-2.24.0/lib/loofah/html5/scrub.rb:
  3: require "cgi"

vendor/ruby/3.4.0/gems/minitest-ci-3.4.0/lib/minitest/ci_plugin.rb:
  2: require 'cgi'

vendor/ruby/3.4.0/gems/net-http-persistent-4.0.5/lib/net/http/persistent.rb:
  3: require 'cgi' # for escaping

vendor/ruby/3.4.0/gems/prism-1.3.0/lib/prism/dot_visitor.rb:
  9: require "cgi"

vendor/ruby/3.4.0/gems/rack-3.1.8/lib/rack/mock_response.rb:
  3: require 'cgi/cookie'

vendor/ruby/3.4.0/gems/rdoc-6.9.1/lib/rdoc/code_object/method_attr.rb:
  292:     require 'cgi/util'

vendor/ruby/3.4.0/gems/rdoc-6.9.1/lib/rdoc/code_object/context/section.rb:
  2: require 'cgi/util'

vendor/ruby/3.4.0/gems/rdoc-6.9.1/lib/rdoc/markup/to_html.rb:
  2: require 'cgi/util'

vendor/ruby/3.4.0/gems/rdoc-6.9.1/lib/rdoc/markup/to_label.rb:
  2: require 'cgi/util'

vendor/ruby/3.4.0/gems/rouge-4.5.1/lib/rouge/lexer.rb:
  6: require 'cgi'

vendor/ruby/3.4.0/gems/rouge-4.5.1/lib/rouge/formatters/html_legacy.rb:
  5: require 'cgi'

vendor/ruby/3.4.0/gems/rubocop-1.72.2/lib/rubocop/formatter/html_formatter.rb:
  3: require 'cgi'

vendor/ruby/3.4.0/gems/ruby-lsp-0.23.17/lib/ruby_lsp/internal.rb:
  17: require "cgi"

vendor/ruby/3.4.0/gems/sidekiq-8.0.2/lib/sidekiq/web/helpers.rb:
  5: require "cgi"

vendor/ruby/3.4.0/gems/w3c_validators-1.3.7/lib/w3c_validators/validator.rb:
  1: require 'cgi'
</p>
</details> 

For a handful I have already made PRs and I do want to keep chugging along and doing a few when I have time but it will take a while. I don't think there's something to reliably keep track

byroot replied:
> I don't think there's something to reliably keep track

That why maybe we should just not add it ourself, and see ruby-head CI fail?

Earlopain replied:
If that is prefered. I would guess it would be for quite a long time though until everything is actually released. But, you are better of to just ignore the warning, without adding it to the gemfile. I removed it.

zzak replied:
Azure is deprecated (and should be removed in 8.1) so I would ignore that.

Earlopain committed:
Only load from `cgi` what is required

In Ruby 3.5 most of the `cgi` gem will be removed (bundled gem). Only the various escape/unescape methods will be retained by default,
which is luckily the only ones that rails actually uses.

Practically, requiring `cgi` will still work, it just emits a warning (and other things like `CGI::Cookie` obviously won't be available).

On older versions, `require "cgi/util"` is needed because the unescape* methods don't work otherwise.

Rails itself only uses escape/unescape functions.

https://bugs.ruby-lang.org/issues/21258

byroot commented:
Ok, so I wasn't crazy, we do end up requiring `cgi/cookie`, but it's through `rack`:

```
/usr/local/lib/ruby/3.5.0+0/bundled_gems.rb:59: warning: CGI library is removed from Ruby 3.5. Please use cgi/escape instead for CGI.escape and CGI.unescape features.
If you need to use the full features of CGI library, Please install cgi gem.
rake aborted!
LoadError: cannot load such file -- cgi/cookie (LoadError)
/rails/actionpack/lib/action_controller/metal/strong_parameters.rb:11:in '<top (required)>'
/rails/actionpack/lib/action_controller.rb:8:in '<top (required)>'
```

Earlopain commented:
Yeah, that's the PR I was linking to https://github.com/rack/rack/pull/2327

Earlopain commented:
@byroot would you be ok to backport to 8.0? I can open a PR.

byroot commented:
For now I backported adding `cgi` as a dependency. I think it's the simpler path for now. Once we got new release of rack then we can consider doing a cleaner fix.
